lib/kafka/protocol/txn_offset_commit_response.rb

```diff
@@ -1,17 +1,44 @@
 module Kafka
   module Protocol
     class TxnOffsetCommitResponse
+      class PartitionError
+        attr_reader :partition, :error_code
 
-      attr_reader :error_code
+        def initialize(partition:, error_code:)
+          @partition = partition
+          @error_code = error_code
+        end
+      end
+
+      class TopicPartitionsError
+        attr_reader :topic, :partitions
+
+        def initialize(topic:, partitions:)
+          @topic = topic
+          @partitions = partitions
+        end
+      end
+
+      attr_reader :errors
 
-      def initialize(error_code:)
-        @error_code = error_code
+      def initialize(errors:)
+        @errors = errors
       end
 
       def self.decode(decoder)
         _throttle_time_ms = decoder.int32
-        error_code = decoder.int16
-        new(error_code: error_code)
+        errors = decoder.array do
+          TopicPartitionsError.new(
+            topic: decoder.string,
+            partitions: decoder.array do
+              PartitionError.new(
+                partition: decoder.int32,
+                error_code: decoder.int16
+              )
+            end
+          )
+        end
+        new(errors: errors)
```

Comment:
Could you explain what this line is doing in more depth? Which class does this `new` operator match?

Reply:
we decode response written on Kafka protocol and give this data to transaction manager. Transaction manager analyses it:
```
        send_response.errors.each do |tp|
          tp.partitions.each do |partition|
            Protocol.handle_error(partition.error_code)
          end
        end
```
and verifies if there is no error codes (error code 0 means success)